File: spring-rest-docs/src/main/java/com/example/restdocs/member/presentation/MemberController.java
package com.example.restdocs.member.presentation;


import com.example.restdocs.member.presentation.dto.MemberModificationRequest;
import com.example.restdocs.member.domain.MemberRepository;
import com.example.restdocs.member.presentation.dto.MemberResponse;
import com.example.restdocs.member.presentation.dto.MemberSignUpRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.PageableDefault;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;

@RestController
@RequestMapping("/api/members")
@RequiredArgsConstructor
@Transactional
public class MemberController {

    private final MemberRepository memberRepository;

    @GetMapping("/{id}")
    public MemberResponse getMember(@PathVariable Long id){
        return new MemberResponse(memberRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 회원입니다.")));
    }

    @PostMapping
    public void createMember(@RequestBody @Valid MemberSignUpRequest dto){
        memberRepository.save(dto.toEntity());
    }

    @PatchMapping("/{id}")
    public void modify(@PathVariable Long id ,@RequestBody @Valid MemberModificationRequest dto){
        memberRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 회원입니다."))
                .modify(dto.getAge());
    }

    @GetMapping
    public Page<MemberResponse> getMembers(
            @PageableDefault(sort = "id",direction = Sort.Direction.DESC)Pageable pageable){
        return memberRepository.findAll(pageable).map(MemberResponse::new);
    }

}

File: webflux-coroutine/server/src/main/kotlin/com/example/server/controller/MemberController.kt
package com.example.server.controller

import com.example.server.dao.Member
import com.example.server.dto.MemberRequest
import com.example.server.service.MemberService
import kotlinx.coroutines.flow.Flow
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/member")
class MemberController(
    private val memberService: MemberService
) {

    @PostMapping
    suspend fun createMember(@RequestBody memberRequest: MemberRequest): Member {
        return memberService.createMember(memberRequest)
    }

    @GetMapping("/{id}")
    suspend fun findOne(@PathVariable id: Long): Member {
        return memberService.findMember(id)
    }

    @GetMapping("/with-team/{id}")
    suspend fun findOneWithTeam(@PathVariable id: Long): Member {
        return memberService.findMemberWithTeam(id)
    }

    @PatchMapping("/{id}")
    suspend fun updateMember(@PathVariable id: Long, @RequestBody memberRequest: MemberRequest) {
        return memberService.updateMember(id, memberRequest)
    }

    @DeleteMapping("/{id}")
    suspend fun delete(@PathVariable id: Long) {
        return memberService.delete(id)
    }

    @GetMapping
    suspend fun findMembers(): Flow<Member> {
        return memberService.findMembers()
    }

    @GetMapping("/with-team")
    suspend fun findMembersWithTeam(): Flow<Member> {
        return memberService.findMembersWithTeam()
    }
}
